File: sponge-nashorn/examples/script/js/unordered_rules.js
/**
 * Sponge Knowledge base
 * Using unordered rules
 */

var AtomicInteger = java.util.concurrent.atomic.AtomicInteger;
var Deduplication = org.openksavi.sponge.core.library.Deduplication;

function onInit() {
    // Variables for assertions only
    EPS.setVariable("hardwareFailureJavaCount", new AtomicInteger(0));
    EPS.setVariable("hardwareFailureScriptCount", new AtomicInteger(0));
    EPS.setVariable("sameSourceFirstFireCount", new AtomicInteger(0));
}

var FirstRule = Java.extend(Rule, {
    onConfigure: function(self) {
        self.events = ["filesystemFailure", "diskFailure"]; self.ordered = false;
        self.addAllConditions(function(rule, event) { return rule.firstEvent.get("source") == event.get("source"); })
        self.addAllConditions(function(rule, event) { 
            return Duration.between(rule.firstEvent.time, event.time).seconds <= 2;
        });
        self.duration = Duration.ofSeconds(5);
    },
    onRun: function(self, event) {
        self.logger.debug("Running rule for events: {}", self.eventSequence);
        EPS.getVariable("sameSourceFirstFireCount").incrementAndGet();
        EPS.event("alarm").set("source", self.firstEvent.get("source")).send();
    }
});

var SameSourceAllRule = Java.extend(Rule, {
    onConfigure: function(self) {
        // Events specified with aliases (e1 and e2)
        self.events = ["filesystemFailure e1", "diskFailure e2 :all"]; self.ordered = false;
        self.addConditions("e1", this.severityCondition);
        self.addConditions("e2", this.severityCondition, function(rule, event) {
            return event.get("source") == rule.firstEvent.get("source") &&
                Duration.between(rule.firstEvent.time, event.time).seconds <= 4;
        });
        self.duration = Duration.ofSeconds(8);
    },
    onRun: function(self, event) {
        self.logger.info("Monitoring log [{}]: Critical failure in {}! Events: {}", event.time, event.get("source"),
                self.eventSequence);
        EPS.getVariable("hardwareFailureScriptCount").incrementAndGet();
    },
    severityCondition: function(self, event) {
        return parseInt(event.get("severity")) > 5;
    }
});

var AlarmFilter = Java.extend(Filter, {
    onConfigure: function(self) {
        self.event = "alarm";
    },
    onInit: function(self) {
        // There is some magic required here because of the limitations in JavaScript support.
        self.target = new function() {
            this.deduplication = new Deduplication("source");
        }
        self.target.deduplication.cacheBuilder.expireAfterWrite(2, TimeUnit.SECONDS);
    },
    onAccept: function(self, event) {
        return self.target.deduplication.onAccept(event);
    }
});

var Alarm = Java.extend(Trigger, {
    onConfigure: function(self) {
        self.event = "alarm";
    },
    onRun: function(self, event) {
        self.logger.debug("Received alarm from {}", event.get("source"));
    }
});

function onLoad() {
    EPS.enableJava(org.openksavi.sponge.examples.SameSourceJavaUnorderedRule.class);
}

function onStartup() {
    EPS.event("diskFailure").set("severity", 10).set("source", "server1").send();
    EPS.event("diskFailure").set("severity", 10).set("source", "server2").send();
    EPS.event("diskFailure").set("severity", 8).set("source", "server1").send();
    EPS.event("diskFailure").set("severity", 8).set("source", "server1").send();
    EPS.event("filesystemFailure").set("severity", 8).set("source", "server1").send();
    EPS.event("filesystemFailure").set("severity", 6).set("source", "server1").send();
    EPS.event("diskFailure").set("severity", 6).set("source", "server1").send();
}

File: sponge-jython/examples/script/py/unordered_rules.py
"""
Sponge Knowledge Base
Using unordered rules
"""

from java.util.concurrent.atomic import AtomicInteger
from org.openksavi.sponge.examples import SameSourceJavaUnorderedRule
from org.openksavi.sponge.core.library import Deduplication

def onInit():
    # Variables for assertions only
    sponge.setVariable("hardwareFailureJavaCount", AtomicInteger(0))
    sponge.setVariable("hardwareFailureScriptCount", AtomicInteger(0))
    sponge.setVariable("sameSourceFirstFireCount", AtomicInteger(0))


class FirstRule(Rule):
    def onConfigure(self):
        self.withEvents(["filesystemFailure", "diskFailure"]).withOrdered(False)
        self.withAllConditions([
            lambda rule, event: rule.firstEvent.get("source") == event.get("source"),
            lambda rule, event:Duration.between(rule.firstEvent.time, event.time).seconds <= 2
        ])
        self.withDuration(Duration.ofSeconds(5))
    def onRun(self, event):
        self.logger.debug("Running rule for events: {}", self.eventSequence)
        sponge.getVariable("sameSourceFirstFireCount").incrementAndGet()
        sponge.event("alarm").set("source", self.firstEvent.get("source")).send()

class SameSourceAllRule(Rule):
    def onConfigure(self):
        self.withEvents(["filesystemFailure e1", "diskFailure e2 :all"]).withOrdered(False)
        self.withCondition("e1", self.severityCondition)
        self.withConditions("e2", [self.severityCondition, self.diskFailureSourceCondition])
        self.withDuration(Duration.ofSeconds(5))
    def onRun(self, event):
        self.logger.info("Monitoring log [{}]: Critical failure in {}! Events: {}", event.time, event.get("source"),
                         self.eventSequence)
        sponge.getVariable("hardwareFailureScriptCount").incrementAndGet()
    def severityCondition(self, event):
        return int(event.get("severity")) > 5
    def diskFailureSourceCondition(self, event):
        # Both events have to have the same source
        return event.get("source") == self.firstEvent.get("source") and \
            Duration.between(self.firstEvent.time, event.time).seconds <= 4

class AlarmFilter(Filter):
    def onConfigure(self):
        self.withEvent("alarm")
    def onInit(self):
        self.deduplication = Deduplication("source")
        self.deduplication.cacheBuilder.expireAfterWrite(2, TimeUnit.SECONDS)
    def onAccept(self, event):
        return self.deduplication.onAccept(event)

class Alarm(Trigger):
    def onConfigure(self):
        self.withEvent("alarm")
    def onRun(self, event):
        self.logger.debug("Received alarm from {}", event.get("source"))

def onLoad():
    sponge.enableJava(SameSourceJavaUnorderedRule)

def onStartup():
    sponge.event("diskFailure").set("severity", 10).set("source", "server1").send()
    sponge.event("diskFailure").set("severity", 10).set("source", "server2").send()
    sponge.event("diskFailure").set("severity", 8).set("source", "server1").send()
    sponge.event("diskFailure").set("severity", 8).set("source", "server1").send()
    sponge.event("filesystemFailure").set("severity", 8).set("source", "server1").send()
    sponge.event("filesystemFailure").set("severity", 6).set("source", "server1").send()
    sponge.event("diskFailure").set("severity", 6).set("source", "server1").send()

